### PR TITLE
ICP API prop test types

### PR DESCRIPTION
**Pull request by bdemann**
-   [x] Breaking changes enumerated in [the release issue](https://github.com/demergent-labs/azle/issues/2053)
-   [x] New documentation enumerated in [the release issue](https://github.com/demergent-labs/azle/issues/2053)
-   [x] Code has been declaratized
-   [x] Error handling beautiful (no unwraps or expects etc)
-   [x] Code tested thoroughly

### CODE REVIEW

File: tests/property/ic_api/cycles/test/tests.ts [hunk 1]
```diff
@@ -78,6 +78,38 @@ export function getTests(): Test {
                 defaultPropTestParams()
             );
         });
+
```
Comment (bdemann) [inline tests/property/ic_api/cycles/test/tests.ts:81]: One

File: tests/property/ic_api/cycles_burn/test/tests.ts [hunk 2]
```diff
@@ -42,5 +42,17 @@ export function getTests(): Test {
                 defaultPropTestParams()
             );
         });
+
+        it('asserts canisterBalance static and runtime types', async () => {
+            const actor = await getCanisterActor<Actor>('canister');
+            const result = await actor.assertCanisterBalanceTypes();
```
Comment (bdemann) [inline tests/property/ic_api/cycles_burn/test/tests.ts:48]: two

File: tests/property/ic_api/id/test/tests.ts [hunk 3]
```diff
@@ -13,6 +13,12 @@ import { _SERVICE as Actor } from './dfx_generated/canister/canister.did';
 
 export function getTests(): Test {
     return () => {
+        it('asserts id static and runtime types', async () => {
```
Comment (bdemann) [inline tests/property/ic_api/id/test/tests.ts:16]: three

File: tests/property/ic_api/instruction_counter/test/tests.ts [hunk 4]
```diff
@@ -87,6 +87,12 @@ export function getTests(): Test {
                 defaultPropTestParams()
             );
         });
+
+        it('asserts instructionCounter static and runtime types', async () => {
+            const actor = await getCanisterActor<Actor>('canister');
+            const result = await actor.assertTypes();
```
Comment (bdemann) [inline tests/property/ic_api/instruction_counter/test/tests.ts:93]: four

File: tests/property/ic_api/is_controller/test/tests.ts [hunk 5]
```diff
@@ -149,5 +149,13 @@ export function getTests(): Test {
                 defaultPropTestParams()
             );
         });
+
+        it('asserts isController static and runtime types', async () => {
+            const actor = await getCanisterActor<Actor>('canister');
+            const result = await actor.assertTypes(
```
Comment (bdemann) [inline tests/property/ic_api/is_controller/test/tests.ts:155]: five

File: tests/property/ic_api/performance_counter/test/tests.ts [hunk 6]
```diff
@@ -168,6 +168,14 @@ export function getTests(): Test {
                 defaultPropTestParams()
             );
         });
+
+        it('asserts performanceCounter static and runtime types', async () => {
+            const actor = await getCanisterActor<Actor>('canister');
+            const result0 = await actor.assertTypes(0);
```
Comment (bdemann) [inline tests/property/ic_api/performance_counter/test/tests.ts:174]: six

File: tests/property/ic_api/reject_code/test/tests.ts [hunk 7]
```diff
@@ -22,5 +22,11 @@ export function getTests(): Test {
             const result = await callerCanister.getRejectNoError();
             expect(result).toEqual({ NoError: null });
         });
+
+        it('asserts rejectCode static and runtime types', async () => {
+            const callerCanister = await getCanisterActor<Actor>('caller');
+            const result = await callerCanister.assertTypes();
```
Comment (bdemann) [inline tests/property/ic_api/reject_code/test/tests.ts:28]: seven

File: tests/property/ic_api/reject_message/test/tests.ts [hunk 8]
```diff
@@ -22,5 +22,11 @@ export function getTests(): Test {
                 defaultPropTestParams()
             );
         });
+
+        it('asserts rejectMessage static and runtime types', async () => {
+            const callerCanister = await getCanisterActor<Actor>('caller');
+            const result = await callerCanister.assertTypes('');
```
Comment (bdemann) [inline tests/property/ic_api/reject_message/test/tests.ts:28]: eight

File: tests/property/ic_api/time/test/tests.ts [hunk 9]
```diff
@@ -106,6 +106,12 @@ export function getTests(): Test {
                 defaultPropTestParams()
             );
         });
+
+        it('asserts time static and runtime types', async () => {
+            const actor = await getCanisterActor<Actor>('canister');
+            const result = await actor.assertTypes();
```
Comment (bdemann) [inline tests/property/ic_api/time/test/tests.ts:112]: nine

File: tests/property/ic_api/reject_code/src/caller/index.ts [hunk 10]
```diff
@@ -1,37 +1,96 @@
...
+        'DestinationInvalid',
+        'CanisterReject',
+        'CanisterError',
+        'Unknown'
+    ];
+    const keys = Object.keys(code);
+    return (
+        keys.length === 1 &&
+        validKeys.includes(keys[0]) &&
+        (code as any)[keys[0]] === null
```
Comment (lastmjs) [inline tests/property/ic_api/reject_code/src/caller/index.ts:92]: Can we try to improve this type?

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: tests/property/ic_api/certified_data/src/index.ts [hunk 11]
```diff
@@ -93,4 +94,39 @@ export default class {
     setDataCertificateInQuery(): void {
         setCertifiedData(new Uint8Array([3]));
     }
+
+    @query([], IDL.Bool)
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/certified_data/src/index.ts:98]: Yes sorry, I thought I had done that but the get got lost in the long name, good catch

File: tests/property/ic_api/certified_data/test/tests.ts [hunk 12]
```diff
@@ -297,6 +297,27 @@ export function getTests(): Test {
...
         });
+
+        it('asserts dataCertificate static and runtime types in query', async () => {
+            const actor = await getCanisterActor<Actor>(CANISTER_NAME);
+            expect(await actor.assertGetDataCertificateTypesInQuery()).toBe(
+                true
+            );
+        });
+
+        it('verifies that dataCertificate returns undefined in update', async () => {
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/certified_data/test/tests.ts:308]: done

File: tests/property/ic_api/certified_data/test/tests.ts [hunk 13]
```diff
@@ -297,6 +297,27 @@ export function getTests(): Test {
...
+            const actor = await getCanisterActor<Actor>(CANISTER_NAME);
+            expect(await actor.assertGetDataCertificateTypesInUpdate()).toBe(
+                true
+            );
+        });
+
+        it('verifies that setCertifiedData returns undefined and takes a Uint8Array', async () => {
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/certified_data/test/tests.ts:315]: done

File: tests/property/ic_api/cycles/src/intermediary/index.ts [hunk 14]
```diff
@@ -58,6 +66,48 @@ export default class {
...
+
+    @query([], IDL.Bool)
+    assertCanisterBalanceTypes(): boolean {
+        type _AssertReturnType = AssertType<
+            NotAnyAndExact<ReturnType<typeof canisterBalance>, bigint>
+        >;
+        return typeof canisterBalance() === 'bigint';
+    }
+
+    @update([IDL.Nat64], IDL.Bool)
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/cycles/src/intermediary/index.ts:78]: Sure

File: tests/property/ic_api/arg_data_raw/test/tests.ts [hunk 15]
```diff
@@ -76,6 +76,8 @@ export function getTests(): Test {
                             console.info('Update method iteration:', i);
                             await testCanisterMethod(updateArgDef, 'Update');
                         }
+
+                        expect(await actor.assertTypes()).toBe(true);
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/arg_data_raw/test/tests.ts]: done

File: tests/property/ic_api/reject_code/src/caller/index.ts (hunk 10 above)
Comment (bdemann) [reply tests/property/ic_api/reject_code/src/caller/index.ts:92]: I have this update, I think its essentially the same... but more typeish I guess...

**Review by lastmjs — APPROVED** [review]

**Review by lastmjs — APPROVED** [review]

**Pull request merged**